libs/core/algorithms/include/hpx/parallel/datapar/adjacent_difference.hpp

```diff
@@ -0,0 +1,73 @@
+//  Copyright (c) 2021 Srinivas Yadav
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+
+#if defined(HPX_HAVE_DATAPAR)
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/execution/traits/is_execution_policy.hpp>
+#include <hpx/functional/tag_dispatch.hpp>
+#include <hpx/iterator_support/zip_iterator.hpp>
+#include <hpx/parallel/algorithms/detail/adjacent_difference.hpp>
+#include <hpx/parallel/datapar/iterator_helpers.hpp>
+#include <hpx/parallel/datapar/loop.hpp>
+#include <hpx/parallel/datapar/zip_iterator.hpp>
+#include <hpx/parallel/util/result_types.hpp>
+
+#include <cstddef>
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename ExPolicy>
+    struct datapar_adjacent_difference
+    {
+        template <typename InIter, typename OutIter, typename Op>
+        static inline OutIter call(
+            InIter first, InIter last, OutIter dest, Op&& op)
+        {
+            if (first == last)
+                return dest;
+            auto count = std::distance(first, last) - 1;
+
+            InIter prev = first;
+            *dest++ = *first++;
+
+            if (count == 0)
+            {
+                return dest;
+            }
+
+            using hpx::get;
+            using hpx::util::make_zip_iterator;
+            util::loop_n<std::decay_t<ExPolicy>>(
+                make_zip_iterator(first, prev, dest), count, [op](auto&& it) {
+                    get<2>(*it) =
+                        hpx::util::invoke(op, get<0>(*it), get<1>(*it));
```

Comment:
Please use `HPX_INVOKE` instead (should be used everywhere in the implementation).